src/main/java/fi/aalto/cs/apluscourses/ui/base/TreeView.java

```diff
@@ -42,24 +56,38 @@ public TreeView() {
    */
   public void setViewModel(@Nullable BaseTreeViewModel<?> viewModel) {
     if (viewModel != null) {
-      DefaultTreeModel treeModel = new DefaultTreeModel(createNode(viewModel));
-      viewModel.filtered.addListener(treeModel, DefaultTreeModel::reload);
-      setModel(treeModel);
+      synchronized (viewModelLock) {
+        unregisterViewModel();
+        this.viewModel = viewModel;
+        registerViewModel();
+      }
+      update();
     }
     selectedItem = null;
   }
 
-  @NotNull
-  private static DefaultMutableTreeNode createNode(@NotNull Tree tree) {
-    List<? extends Tree> children = tree.getChildren();
-    OurTreeNode node = new OurTreeNode(tree);
-    if (tree instanceof Filterable) {
-      ((Filterable) tree).addVisibilityListener(node);
+  private void unregisterViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.removeCallback(this);
+      }
     }
-    for (Tree subtree : children) {
-      node.add(createNode(subtree));
+  }
+
+  private void registerViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.addListener(this, TreeView::update);
```

Comment:
Could this possibly call `update` (which modifies the UI tree model) outside of the EDT?

Reply:
That's a valid point. It probably could be reasoned out whether or not that method is ever called outside of the EDT but it's better to be sure and add `invokeLater` there somewhere.

What is more, if that method is called outside of EDT, it could, in theory at least, be possible that of two subsequent filterings, the latter would be rewritten by the former (if it happened very soon after the other). That would result in a situation where tree is filtered by a different filter than what option buttons display. But that is a very minor concern, anyways.